feat(editor): enhance section rendering, area with multi-variant, omni-spawn and build optimizations

warmonipa opened the pull request:
## Editor

* **Room ID & Rendering**

  * Implement `RoomIdRenderer` with toolbar toggle.
  * Display first byte of Door ID for Forest Door objects.
  * Add `object_room_id` circle rendering for CCA/Forest areas.
  * Add origin position (0,0,0) display toggle

## Gameplay

* **Event & NPC**

  * Add `RangeCircleRenderer` for EventCollision radius (real-time updates).
  * Add NPC spawn toggle with dynamic ground height via raycast.
  * Add omnispawn functionality to show monsters from all episodes:

## UI

* **Area & Map**

  * Improve area-variant display — use 1-based indexing.
  * Enable EP1 Battle Area maps (Lobby, BA Spaceship, BA Palace).
  * Enhance map designations and multi-area quest support
  * Enhanced camera navigation to preserve user's zoom and viewpoint

* **3D View Enhancements**
  
  * Add world position label in 3D view.
  * Enhance section number ring visibility with increased thickness.

## Assets

* Fix resource paths for LaDimenian / SoDimenian.

## Build

* **Gradle & Kotlin**

  * Fix Gradle task dependencies in `web-assembly-worker`.
  * Change `kotlin.daemon.jvmArgs` from `4T` to `4G`.
* **Webpack**

  * Optimize webpack build performance for event modules.

warmonipa committed:
feat(quest): enhance floor mapping support for multi-segment quests

Commit:
refactor(psolib): extract GameArea enum and simplify area mapping functions

  - Extract GameArea enum to separate file as single source of truth
  - Simplify getAreaIdByMapId/getMapId from 56+ lines to 1-line implementations
  - Add bidirectional lookup with O(1) performance using pre-computed indices
  - Update FloorMapping to include mapId field and standardize naming
  - Enhance MAP_DESIGNATE instruction handling with floorId register reading
  - Update opcodes.yml with complete map_designate register definitions

warmonipa committed:
chord: add comments and rename rawFloorId to floorId

DaanVandenBosch reviewed (changes requested):
I started reviewing this, but it's a lot to get through. Do you think you could split this up into several different, smaller PRs? This will make it much quicker for me to review and your changes will get merged faster.

For example, one commit with small changes such as naming changes to opcodes and entities, one commit for omnispawn, one commit with gradle config changes, and so on.

DaanVandenBosch commented:
I've manually "merged" the Gradle fixes, as this was necessary to get the project working on Linux. Oddly, GitHub actions have no problem with the old config.

warmonipa commented:
> I've manually "merged" the Gradle fixes, as this was necessary to get the project working on Linux. Oddly, GitHub actions have no problem with the old config.

Thanks for the reply!

Just to clarify — I actually closed two earlier MRs because the review process was taking a while, and I wasn’t sure if they were still being looked at. Also, since I hadn’t worked with Kotlin before, that slowed me down a bit.

I’ve now made all the adjustments as requested. Let me know if anything else needs to be updated!

warmonipa commented on web/src/jsMain/kotlin/world/phantasmal/web/questEditor/controllers/EntityListController.kt:
```diff
@@ -16,24 +17,69 @@ class EntityListController(store: QuestEditorStore, private val npcs: Boolean) :
+            }
+
+            // Boss areas: only show current floor's boss NPCs (original logic)
+            if (isBossArea(episode, areaId)) {
```
That makes sense! Passing the whole area instead of just the area ID would simplify things and eliminate the need for the isBossArea function.

warmonipa committed:
Merge branch 'master' into master

warmonipa commented on web/src/jsMain/kotlin/world/phantasmal/web/questEditor/controllers/EntityListController.kt:
```diff
@@ -16,24 +17,69 @@ class EntityListController(store: QuestEditorStore, private val npcs: Boolean) :
+                // For objects, show all objects from any area in all episodes
+                return entityType.areaIds[Episode.I]?.isNotEmpty() == true ||
+                        entityType.areaIds[Episode.II]?.isNotEmpty() == true ||
+                        entityType.areaIds[Episode.IV]?.isNotEmpty() == true
```
I’ve simplified the check to entityType.areaIds.isNotEmpty() as suggested. Additionally, I’ve added a check in the ObjectType constructor to ensure we never put empty lists in the map.

warmonipa commented on web/src/jsMain/kotlin/world/phantasmal/web/questEditor/controllers/EntityListController.kt:
```diff
@@ -16,24 +17,69 @@ class EntityListController(store: QuestEditorStore, private val npcs: Boolean) :
 
-            (entityType.episode == null || entityType.episode == episode) &&
-                    areaId in entityType.areaIds
+            // Always filter out specific enemy NPCs that should be removed
```
I understand your point. I’ll go ahead and remove the comment.

warmonipa committed:
refactor(quest-editor): simplify entity filtering logic

  - Pass AreaModel instead of areaId to avoid isBossArea() linear search
  - Remove duplicate episode/area matching logic (was in 3 places)
  - Remove redundant !entityType.enemy check for Pioneer2 (no enemy NPCs have areaId 0)
  - Simplify ObjectType omnispawn filter to check areaIds.isNotEmpty()
  - Add ObjectType validation to ensure areaIds never contains empty lists

DaanVandenBosch commented:
There's some more things I want to comment on, but it's going to take me some time to get through it all. It's a lot in a single PR. Again, if you could split this up, it would be much easier to see all the relevant changes together and I could get through it faster.

Commit:
chore: remove unnecessary comments and logs

warmonipa commented:
> There's some more things I want to comment on, but it's going to take me some time to get through it all. It's a lot in a single PR. Again, if you could split this up, it would be much easier to see all the relevant changes together and I could get through it faster.

Thanks for taking the time to review this! I get that it’s a long PR which is somewhat connected, so breaking it apart would make it harder to keep things consistent. I’ll add clear notes and explanations to make the review easier.

Commit:
chore: remove unnecessary comments and logs

Commit:
chore: add boss, pioneer2, and lab utility area checks

Commit:
chore(FloorMapping): add note about switching from floor-variant to areaId-variant mapping to fix multi-variant map parsing

Commit:
chore: change variable name `couldBeBbMapDesignate` to `multiVariant`
